napari/utils/progress.py

```diff
@@ -150,3 +150,42 @@ def progrange(*args, **kwargs):
 
     """
     return progress(range(*args), **kwargs)
+
+
+class cancelable_progress(progress):
+    def __init__(
+        self,
+        iterable: Optional[Iterable] = None,
+        desc: Optional[str] = None,
+        total: Optional[int] = None,
+        nest_under: Optional['progress'] = None,
+        cancel_callback: Optional[Callable] = None,
+        *args,
+        **kwargs,
+    ) -> None:
+        self.cancel_callback = cancel_callback
+        self.is_canceled = False
+
+        super().__init__(iterable, desc, total, nest_under, *args, **kwargs)
+
+    def __iter__(self) -> Iterator:
+        plain_iterator = super().__iter__()
+
+        # Wrap the tqdm iterator to allow e.g. cancellation
+        class Wrapped(Iterator):
+            def __iter__(wrapped):
+                return wrapped
+
+            def __next__(wrapped):
+                if self.is_canceled:
+                    raise StopIteration
+                return next(plain_iterator)
+
+        return Wrapped()
```

Comment:
Maybe takewhile may be used? https://docs.python.org/3/library/itertools.html#itertools.takewhile

Reply:
Great idea, fixed with 6b4ddab